### ISSUE TITLE
Certain videos switching language tracks without user interaction...

### ISSUE THREAD

**Issue by attaboyBrad** [fer0n/Unwatched#35]
A particular Gerald Undone channel video (https://youtu.be/4OcP9XPb2kI) keeps changing languages mid-stream and I can't figure out how to change it back.  It plays with the default English on Youtube web.  Occasionally if I scrub it will restart playing in English for a second or two in Unwatched but then it changes itself back to Spanish without any interaction on my part.

I see on [this thread](https://www.reddit.com/r/youtube/comments/1ddv6sa/incorrect_audio_tracks/) that multiple audio tracks are a new feature, so maybe it's an issue with how the channel set their own audio tracks.

[Here's a screen recording](https://youtu.be/DUCrvlgL01g) demonstrating the issue.

Having now watched the video... it was a pretty useless video.  Maybe Unwatched was trying to spare me those minutes wasted.

**Comment by attaboyBrad** [comment]
Same thing is happening in [this video](https://youtu.be/MCTyhADd2L4) from another channel, so it's not a one-off issue or channel specific.

![ScreenCap 2024-12-10  Google Chrome  The Strangest Thing Just Happened - YouTube  000558](https://github.com/user-attachments/assets/1f1cc1ef-0909-4fbf-9d85-584e4dc2d31f)

The setting to switch audio tracks in the attached screenshot shows up in Safari and Chrome, but not in Unwatched.

**Comment by attaboyBrad** [comment]
http://www.techmeme.com/241210/p19#a241210p19

This seems likely to be relevant...

**Comment by fer0n** [comment]
Thanks so much for the detailed description and screen recordings :) 

Definitely odd. Did I understand that correctly, the issue is exclusively happening on macOS and only inside the player?

I'm using the YouTube embedded player and for some reason YouTube doesn't appreciate having the embedded player blown up in size on macOS. Pip doesn't work, native fullscreen doesn't work, and neither does the quality selection (it was also stuck on 480p).

I tried a few things to work around this, but the best I was able to achieve was having a higher quality (but broken selection) combined with an "hide Unwatched controls"- fullscreen. So I fear there won't be a lot I can do about this, but I'll take a look at it.

**Comment by attaboyBrad** [comment]
Correct on "only MacOS and only inside the player".  I tried and failed to reproduce the issue on iPadOS.

**Comment by attaboyBrad** [comment]
No idea if this is the same issue, but you might also look into whether or not speeds above 2.0 are working on your end on MacOS.  No idea if you'll be able to fix it, but since it involves the Unwatched UI I suppose you'd want to pull those speed options for MacOS if you can't.  Those speeds seem to work fine on iPad, so maybe it's also to do with whatever modifications you made to the player for MacOS.

Screen recording below

https://github.com/user-attachments/assets/895ec0d1-2861-4faa-82a1-517766595535

**Comment by tisba** [comment]
This issue is pretty wild, in one video it was switching from German to Spanish and in the next one it was Russian (I think) from the beginning. I first thought the youtube channel was highjacked!

**Comment by attaboyBrad** [comment]
I saw that the [Vinegar](https://apps.apple.com/us/app/vinegar-tube-cleaner/id1591303229) Safari extension was recently updated to 'ignore the AI generated audio tracks when selecting the default audio track', so it would seem there is some way to do it--but they're replacing the player so that might be a completely different problem.

**Comment by fer0n** [comment]
I saw the vinegar update as well, not entirely sure what they're doing.

I've played around with this for a bit and the only way I've found so far to avoid this without introducing other issues like the quality being limited to 360p is to run it as a native Mac version, which would be quite a bit of effort to do. I'll see if I can come up with something else.

**Comment by attaboyBrad** [comment]
I noticed this morning on build 180 that a few videos started off with the wrong language but switched back to the correct language after a second or two.  Not sure if it's a change you made or a change Youtube made, but the issue may be resolved--I'll mark it closed in another day or two if I can't reproduce it again.

**Comment by fer0n** [comment]
I haven't changed anything that would address this, it's probably related to YouTube. The only realistic option I've found so far is to make a native Mac version to replace the "Made for iPad" version, but that would probably be quite the task.

**Comment by fer0n** [comment]
@attaboyBrad has the issue reappeared for you?

**Comment by attaboyBrad** [comment]
Yes.  Still happens on the same channels noted previously as of today's latest update.

**Comment by fer0n** [comment]
Thanks for the update.

I'll consider a native Mac version, as that seems to be the only way to get around this.

Don't know how much effort it would take, and Mac is around 4% usage share so it's not the highest priority, especially since there's still tons of other stuff to do. I can't promise if or when I'll get to it, but I'll consider it.

**Comment by fer0n** [comment]
I'm assuming it doesn't matter breaking the YouTube ToS if you're not using the YouTube API, which is probably the case for all of these. Unwatched is using the API for a few things, so I'm not too eager to break the ToS. Maybe it'd still be fine, but I'd rather not find out if I can avoid it.

**Comment by attaboyBrad** [comment]
Not expecting any response here, just adding info in cases it's useful down the road.  This is all on Version 1.5.2 (187).  Haven't seen the 'watch video in browser' update yet, but will keep an eye out.

I found an example of a single channel where one video plays in English and another uses AI dubbing.  I previously understood that this was a feature that was enabled by channel, but this observations suggests otherwise.

Here's the one that played on Unwatched Mac in English: https://www.youtube.com/watch?v=Kx6FVDfxCzc 

Here's the one that played with AI dubbing: https://www.youtube.com/watch?v=SC1mVe7isIA

Maybe that A/B will provide some hints about the root of the issue.

Finally, here's a video that starts playing with AI dubbing, but switches back to the original audiotrack after a few seconds: https://www.youtube.com/watch?v=EHwT6VaAkrc

I had previously only seen videos start in the original audiotrack and switch to AI dubbing.

Nothing jumped out at me in the 'stats for nerds' for any of these, but I'm not particularly familiar with looking at that.

**Comment by Revrean** [comment]
Just chiming in to say it happens too (foreign dubbing) with the new AirPlay HD function when casting to my Apple TV

**Comment by fer0n** [comment]
@Revrean I'm assuming it only happens on Apple TV if that setting is on?

**Comment by Revrean** [comment]
Looks like that's the case. I dunno how to identify which videos have AI dubbing enabled

**Comment by fer0n** [comment]
@attaboyBrad there's a new TestFlight version out now, would you mind double checking if that fixes the language issue for you?

**Comment by attaboyBrad** [comment]
> [@attaboyBrad](https://github.com/attaboyBrad) there's a new TestFlight version out now, would you mind checking if that fixes the language issue for you?

The dubbing appears to be fixed, but the player appears to be locked to 360p and it crashes when I  try to view it in Youtube (though "Watch in Browser" works fine.

**Comment by fer0n** [comment]
Hm that's odd, I'll check what's happening there. Just to be sure, what's the version/build number you were testing it on?

**Comment by attaboyBrad** [comment]
1.6.0 (208).  I saw "Native MacOS Version" in the release notes and opened it right away!  Screencaps attached showing missing resolutions and the evidently low res of the video when set to "auto".

![Image](https://github.com/user-attachments/assets/1b34326a-1634-4939-a89e-1ebd8ebf9df7)

![Image](https://github.com/user-attachments/assets/309faa14-9129-4381-aaf4-cce03ec0dc9d)

**Comment by fer0n** [comment]
Thanks for sharing. My mistake, I have to make a separate upload for macOS now, this is just the iOS version without the macOS workarounds which also explains the crash. I've submitted the native version, but it has to go through app review so it sadly might take a few days.

**Comment by attaboyBrad** [comment]
> Edit: app review is done already. You may have to delete the iOS version first and then select macOS in TestFlight on the top left dropdown.

I'm not seeing any such dropdown... Is there a different TestFlight link for the Mac beta?

**Comment by attaboyBrad** [comment]
'F' not working on MacOS to show/hide the sidebar.

**Comment by fer0n** [comment]
I'll close this issue then as it thankfully seems to be fixed. If there's any other macos related problems feel free to open a new issue :)
